x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java

```diff
@@ -515,6 +499,29 @@ protected void createCheckpoint(ActionListener<Void> listener) {
                 listener.onFailure(new RuntimeException("Failed to retrieve checkpoint", getCheckPointException));
             }));
         }
+
+        private boolean isIrrecoverableFailure(Exception e) {
+            return e instanceof IndexNotFoundException || e instanceof DataFrameConfigurationException;
+        }
+
+        synchronized void handleFailure(Exception e) {
+            if (handleCircuitBreakingException (e)) {
```

Comment:
I think we should have a limit on the number of times we retry that is outside the page reduction equation.

Reply:
agreed but I would do this later, 1st we should improve the initial page size.

Note that `handleCircuitBreakingException` is not unlimited, it returns false if it get below the minimum, should be after 18 reductions with the current settings.

Reply:
nit: there is an extra space in `handleCircuitBreakingException (e)` 

`handleCircuitBreakingException` returns false when `newPageSize < MINIMUM_PAGE_SIZE` which happens after many failures then we have to see `MAX_CONTINUOUS_FAILURES` more failures before the task will stop even thought `handleCircuitBreakingException` has just logged the message "_Search returned with out of memory error after repeated page size reductions to [{0}], unable to continue pivot_"

`handleCircuitBreakingException` returns false if the exception is not a circuit breaker or if it is and `newPageSize < MINIMUM_PAGE_SIZE` this is confusing as we can't tell if there was an error or not so maybe move the exception check here

```suggestion
        CircuitBreakingException circuitBreakingException = getCircuitBreakingException(e);

        if (circuitBreakingException != null) {
            if (handleCircuitBreakingException(circuitBreakingException) == false) {
                stopTaskEtc();
            else {
                return;
            }            
        }
            
```

Reply:
@davidkyle good point. Change: https://github.com/elastic/elasticsearch/pull/41149/commits/fa059e879d3344c24658c81c35f53a200e93bf8a

To set the failure reason, I had to implement it differently to your suggestion. Not very nice to add another abstract method, we should revisit this after the task refactoring done outside of this PR.